### PR TITLE
comms: Add sol-netctl agent API V4

### PR DESCRIPTION
**Pull request by wzhen12**
The agent is necessary in the process of network connection,
that will receive input information from the user and send it
to the network(such as AP and so on).

The API is based on connman D-Bus APIs, it provides register agent,
input login information and so on netctl agent API.

Signed-off-by: Wu Zheng wu.zheng@intel.com

### CODE REVIEW

**Comment by wzhen12** [comment]
The Version 4 of Add sol-netctl agent API.
1. Modify the version 3 errors. (#2303)
2. Split register_agent to register_agent and unregister_agent
3. Rename report_error to request_retry

@ceolin  @barbieri  Please check it. Thanks.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 1]
```diff
@@ -1212,3 +1242,382 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
...
+        agent->release((void *)context->agent_data);
+
+    return 0;
+}
+
+static int
+agent_report_error(sd_bus_message *m, void *data, sd_bus_error *ret_error)
+{
+    struct sol_netctl_service *service;
+    struct ctx *context = (struct ctx *)data;
```
Comment (ceolin) [inline src/lib/comms/sol-netctl-impl-connman.c]: C does not require this cast.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 2]
```diff
@@ -1212,3 +1242,382 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
...
+
+    return 0;
+}
+
+SOL_API int
+sol_netctl_unregister_agent(void)
+{
+    int r;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+    const char *path = CONNMAN_AGENT_PATH;
```
Comment (ceolin) [inline src/lib/comms/sol-netctl-impl-connman.c]: why don't use `CONNMAN_AGENT_PATH` directly ?

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 3]
```diff
@@ -1212,3 +1242,382 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
...
+
+    return -ENOMEM;
+}
+
+static int
+agent_request_input(sd_bus_message *m, void *data, sd_bus_error *ret_error)
+{
+    struct sol_netctl_service *service;
+    const char *path;
+    struct ctx *context = (struct ctx *)data;
```
Comment (ceolin) [inline src/lib/comms/sol-netctl-impl-connman.c]: same here

File: src/lib/comms/include/sol-netctl.h [hunk 4]
```diff
@@ -209,6 +216,60 @@ enum sol_netctl_state {
...
+        const char *error);
+    /**
+     * @brief connection input callback used to inform connection login input
+     *
+     * @param data the user data
+     * @param service the connection login input service
+     * @param inputs the vector of login input type
+     */
+    void (*request_input)(void *data, const struct sol_netctl_service *service,
+        const struct sol_vector *inputs);
```
Comment (ceolin) [inline src/lib/comms/include/sol-netctl.h]: seing the sample this method is not clear to me. This method is called when connman needs info for a specific input right ? In this implementation the user (that have registered the agent) will be called with a list of inputs, then they will need to create another list of inputs with the answers and then call `sol_netctl_request_input` ? Why we simply don't allow the user fill the answer and we use it ?

What I told is right ?

Comment (wzhen12) [inline src/lib/comms/include/sol-netctl.h]: @ceolin 
It is gustavo's suggestion.
If any not clear, @barbieri please figure it out. Thanks.
The reason is that the user can keep the malloc/free the vector data.
If  fill the answer, it is confused to malloc/free the vector data.

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 2 above)
Comment (wzhen12) [inline src/lib/comms/sol-netctl-impl-connman.c]: ok, remove path.

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 3 above)
Comment (wzhen12) [inline src/lib/comms/sol-netctl-impl-connman.c]: ok

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 1 above)
Comment (wzhen12) [inline src/lib/comms/sol-netctl-impl-connman.c]: ok

**Pull request closed without merging**